server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java

```diff
@@ -29,81 +33,122 @@
         var it = this.walManager.readAll();
         while (it.hasNext()) {
             var entry = it.next();
-            var tableName = entry.getTableName();
-            var table = this.tables.computeIfAbsent(tableName,
-                    k -> new MemoryTableImpl(tableName,
-                            this.walManager,
-                            this.storageAccessService,
-                            this.dataRootPath,
-                            this.parquetConfig));
+            var table = this.getTable(entry.getTableName(), true, true);
+            //noinspection ConstantConditions
             table.updateFromWal(entry);
+            if (table.getFirstWalLogId() >= 0) {
+                this.dirtyTables.put(table, "");
+            }
         }
+        this.dumpThread = new DumpThread(DurationStyle.detectAndParse(dumpInterval).toMillis(),
+                DurationStyle.detectAndParse(minNoUpdatePeriod).toMillis());
+        this.dumpThread.start();
     }
 
     public void terminate() {
+        this.dumpThread.terminate();
         this.walManager.terminate();
     }
 
     public List<String> list(String prefix) {
-        return tables.keySet().stream().filter(name -> name.startsWith(prefix)).collect(Collectors.toList());
+        try {
+            return Stream.concat(
+                            this.storageAccessService.list(this.snapshotRootPath + prefix)
+                                    .map(path -> {
+                                        path = path.substring(this.snapshotRootPath.length());
+                                        var index = path.indexOf(PATH_SEPARATOR);
+                                        if (index < 0) {
+                                            return path;
+                                        } else {
+                                            return path.substring(0, index);
+                                        }
+                                    }),
+                            tables.keySet().stream().filter(name -> name.startsWith(prefix)))
+                    .sorted()
+                    .distinct()
+                    .collect(Collectors.toList());
+        } catch (IOException e) {
+            throw new SwProcessException(ErrorType.DATASTORE, "failed to list", e);
+        }
     }
 
     public void update(String tableName,
             TableSchemaDesc schema,
             List<Map<String, Object>> records) {
-        var table = this.tables.computeIfAbsent(tableName,
-                k -> new MemoryTableImpl(tableName,
-                        this.walManager,
-                        this.storageAccessService,
-                        this.dataRootPath,
-                        this.parquetConfig));
+        var table = this.getTable(tableName, true, true);
+        //noinspection ConstantConditions
         table.lock();
         try {
             table.update(schema, records);
+            this.dirtyTables.put(table, "");
```

Comment:
The following scenarios may occur:
After table.update(), these two line executed
https://github.com/star-whale/starwhale/blob/ed0fb83723a98332e0b829fc85867c546de3a812/server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java#L460-L461

And then `this.dirtyTables.put(table, "");` executed

There will be a unexpected `minWalLogIdToRetain`, it may cause unexpected wal file deletion.

Reply:
Done